src/Compilers/Core/Portable/WellKnownTypes.cs

```diff
@@ -496,14 +496,41 @@ internal static class WellKnownTypes
 
         static WellKnownTypes()
         {
+            AssertEnumAndTableInSync();
+
             for (int i = 0; i < s_metadataNames.Length; i++)
             {
                 var name = s_metadataNames[i];
                 var typeId = (WellKnownType)(i + WellKnownType.First);
+                s_nameToTypeIdMap.Add(name, typeId);
+            }
+        }
+
+        [Conditional("DEBUG")]
+        private static void AssertEnumAndTableInSync()
+        {
+            for (int i = 0; i < s_metadataNames.Length; i++)
+            {
+                var name = s_metadataNames[i];
+                var typeId = (WellKnownType)(i + WellKnownType.First);
+
+                string typeIdName;
+                if (typeId == WellKnownType.First)
```

Comment:
If you're going to turn one enum into a particular name, you should assert that they represent the same element, e.g.:
```cs
   Debug.Assert(WellKnownType.Last == WellKnownType.System_IFormatProvider);
```